sklearn/preprocessing/_discretization.py

```diff
@@ -272,15 +323,38 @@ def fit(self, X, y=None, sample_weight=None):
             elif self.strategy == "quantile":
                 quantiles = np.linspace(0, 100, n_bins[jj] + 1)
                 if sample_weight is None:
-                    bin_edges[jj] = np.asarray(np.percentile(column, quantiles))
-                else:
+
                     bin_edges[jj] = np.asarray(
-                        [
-                            _weighted_percentile(column, sample_weight, q)
-                            for q in quantiles
-                        ],
+                        np.percentile(column, quantiles, method=quantile_method),
                         dtype=np.float64,
                     )
+                else:
+                    # TODO: implement averaged_inverted_cdf in
+                    # _weighted_percentile instead of calling it twice and,
+                    # sorting the column twice.
+                    def averaged_weighted_percentile(q):
+                        return (
+                            _weighted_percentile(column, sample_weight, q)
+                            - _weighted_percentile(-column, sample_weight, 100 - q)
+                        ) / 2
```

Comment:
I think should be moved and renamed to its own private function `sklearn.utils.stats._averaged_weighted_percentile` and we should add a unittest in `sklearn/utils/tests/test_stats.py` for instance that:

- it returns the same value as `np.median` when called with unit weights on an even number of data points,
- it returns the same value as `np.percentile` with strategy `"averaged_inverted_cdf"` when called with unit weights on random data,
- that it agrees with `_weighted_percentile` in cases where the return value is not interpolated between two observations, for instance:

```python
data = [0, 1, 2]
sample_weight = [5, 1, 5]
q = 50
```